config/dynamic.c

```diff
@@ -16,11 +16,467 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_cycles.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
 #include "gatekeeper_config.h"
+#include "gatekeeper_launch.h"
+
+/* TODO Get the install-path via Makefile. */
+#define LUA_DY_BASE_DIR     "./lua"
+#define DYNAMIC_CONFIG_FILE "dy.lua"
+
+const char *CLIENT_PROC_ERROR =
+	"function reply() return -1, \"Dynamic Configuration cannot process the request. end\"\n";
+
+const uint16_t LUA_MSG_MAX_LEN = 65535;
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, size_t reply_len)
+{
+	int send_size;
+
+	/* The first two bytes: the length of the message. */
+	send_size = send(conn_fd, (uint16_t *)&reply_len, sizeof(uint16_t), 0);
+	if (send_size != 2) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to send reply message length (%lu)\n",
+			reply_len);
+		return -1;
+	}
+
+	/* Sending the message. */
+	send_size = send(conn_fd, reply_msg, strlen(reply_msg), 0);
+	if (send_size != (int)reply_len) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to send reply message (expected len = %lu, sended len = %d) - %s\n",
+			reply_len, send_size, reply_msg);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+	const char *msg, int msg_len, struct dynamic_config *dy_conf)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	int reply_len;
+	char reply_msg[LUA_MSG_MAX_LEN + 1];
```

Comment:
1. I think we can make the Lua function ``dy_cmd_proc()`` return the reply string. However, the reply string will be generated dynamically by each functional block, so we may have to allocate buffer to hold the reply message, and free it after calling ``dy_cmd_proc()``. By allocating the message buffer in this way may be easier?

2. This should depend on the implementation of (1).